File: src/manager/driver.ts
// The ManagerDriver (MIGRATION-CODEX.md §4): one manager turn, now over a long-lived Codex thread
// instead of an Anthropic createMessage loop. Pop an event → prepend the volatile context header →
// runStreamed → stream items: the final `agent_message` is the manager's reply to the owner (honoring
// the NO_REPLY sentinel); `mcp_tool_call` is internal (memory/orchestration) and only logged; `reasoning`
// is private and never delivered. The async worker model is unchanged: subagent_start (an MCP tool)
// returns immediately, the turn ends with an ack, and the worker's completion re-enters as an event.
//
// Durability is simpler than the Anthropic loop: Codex owns the thread's rollout on disk and runs its
// own compaction, so we keep no ModelMessage transcript — only the thread id, to resume after a
// restart. The same thread object is reused across turns in one process; resume() is for cold wake.

import type { ThreadEvent, UserInput } from "@openai/codex-sdk";

import type { ManagerThread, ManagerThreadFactory } from "./managerCodex.js";
import { friendlyError } from "../workers/runner.js";
import { logger } from "../logger.js";

/** Delivers the manager's user-facing text to the owner's chat (wraps Telegram sendMessage). */
export type DeliverFn = (chatId: number, text: string) => Promise<void>;

/** A turn emits this to absorb an event without messaging the owner (reused from the v0.2 loop).
 *  Suppresses the WHOLE message wherever it appears on its own line, so private reasoning that leads
 *  up to a decision to stay silent is never delivered. */
export const NO_REPLY = "NO_REPLY";

/** What opens a turn: an owner message or rendered worker event, optionally with an owner-sent image. */
export interface TurnInput {
  text: string;
  /** Local path to an owner-sent image (view_image is on); becomes a `local_image` input. */
  imagePath?: string;
}

export interface ManagerUsage {
  inputTokens: number;
  outputTokens: number;
  cachedInputTokens: number;
  reasoningTokens: number;
}

/** A normalized conversation entry for the (off-by-default) Inspector — reconstructed from the Codex
 *  item stream since there is no longer a ModelMessage transcript. */
export interface ConvBlock {
  type: string;
  [key: string]: unknown;
}
export interface ConvMessage {
  role: "user" | "assistant";
  content: ConvBlock[];
}

export interface RunTurnOpts {
  /** Token usage for this turn (from turn.completed). */
  onUsage?: (usage: ManagerUsage) => void;
  /** Observability sink for the reconstructed conversation log (Inspector). */
  onConversation?: (message: ConvMessage) => void;
  /** Host-side delivery gate. Conversation is still recorded; owner-visible sends can be suppressed. */
  allowReply?: () => boolean;
}

export interface ManagerDriverDeps {
  factory: ManagerThreadFactory;
  deliver: DeliverFn;
  /** Builds the volatile per-turn header (core memory + index), prepended to each event's input. */
  buildContextHeader: () => string;
}

export interface ManagerDriver {
  runTurn(input: TurnInput, chatId: number, opts?: RunTurnOpts): Promise<void>;
  /** The current manager thread id, for snapshotting (undefined before the first turn). */
  threadId(): string | undefined;
  /** Seed the resume id from a restored snapshot (used before the first turn). */
  adoptThreadId(id: string | undefined): void;
  /** /new: drop the thread so the next turn starts fresh (working context cleared, memory kept). */
  reset(): void;
}

export function createManagerDriver(deps: ManagerDriverDeps): ManagerDriver {
  let thread: ManagerThread | null = null;
  let pendingResumeId: string | undefined;
  let currentThreadId: string | undefined;

  function ensureThread(): ManagerThread {
    if (!thread) {
      thread = pendingResumeId ? deps.factory.resume(pendingResumeId) : deps.factory.start();
    }
    return thread;
  }

  return {
    threadId: () => currentThreadId,
    adoptThreadId(id) {
      pendingResumeId = id;
      currentThreadId = id;
    },
    reset() {
      thread = null;
      pendingResumeId = undefined;
      currentThreadId = undefined;
    },

    async runTurn(input, chatId, opts) {
      const t = ensureThread();
      const header = deps.buildContextHeader();
      const text = header ? `${header}\n\n---\n\n${input.text}` : input.text;
      const codexInput: UserInput[] = [{ type: "text", text }];
      if (input.imagePath) codexInput.push({ type: "local_image", path: input.imagePath });

      opts?.onConversation?.({
        role: "user",
        content: [
          { type: "text", text: input.text },
          ...(input.imagePath ? [{ type: "image", path: input.imagePath }] : []),
        ],
      });

      let failure: string | undefined;
      let finalReply: string | undefined;
      try {
        const { events } = await t.runStreamed(codexInput);
        for await (const event of events) {
          handleEvent(
            event,
            opts,
            (f) => (failure = f),
            (text) => (finalReply = text),
          );
        }
      } catch (err) {
        failure = (err as Error).message;
      }

      // Capture the thread id once the turn has started it; resume is no longer needed this process.
      if (t.id) {
        currentThreadId = t.id;
        pendingResumeId = undefined;
      }

      if (failure) {
        await deps.deliver(chatId, friendlyError(failure));
        return;
      }

      if (finalReply !== undefined) {
        const reply = applyNoReply(finalReply);
        if (reply && (opts?.allowReply?.() ?? true)) await deps.deliver(chatId, reply);
      }
    },
  };
}

function handleEvent(
  event: ThreadEvent,
  opts: RunTurnOpts | undefined,
  setFailure: (f: string) => void,
  setFinalReply: (text: string) => void,
): void {
  switch (event.type) {
    case "item.completed": {
      const item = event.item;
      if (item.type === "agent_message") {
        opts?.onConversation?.({ role: "assistant", content: [{ type: "text", text: item.text }] });
        // The SDK's buffered `run()` treats the last completed agent message as `finalResponse`.
        // Keep streaming observability, but apply the same owner-delivery rule to avoid progress
        // chatter becoming many Telegram messages from a single turn.
        setFinalReply(item.text);
      } else if (item.type === "mcp_tool_call") {
        logger.debug("Manager tool call", { server: item.server, tool: item.tool, status: item.status });
        opts?.onConversation?.({
          role: "assistant",
          content: [
            {
              type: "tool_use",
              name: `${item.server}.${item.tool}`,
              input: item.arguments,
              status: item.status,
              ...(item.error?.message ? { error: item.error.message } : {}),
            },
          ],
        });
        const resultText = item.error?.message ? `error: ${item.error.message}` : mcpResultText(item.result);
        if (resultText !== undefined) {
          opts?.onConversation?.({ role: "user", content: [{ type: "tool_result", content: resultText }] });
        }
      } else if (item.type === "reasoning") {
        // Private: recorded for the Inspector, never delivered to the owner.
        opts?.onConversation?.({ role: "assistant", content: [{ type: "thinking", thinking: item.text }] });
      } else if (item.type === "error") {
        logger.warn("Manager turn surfaced an error item", { message: item.message });
      }
      break;
    }
    case "turn.completed":
      opts?.onUsage?.({
        inputTokens: event.usage.input_tokens,
        outputTokens: event.usage.output_tokens,
        cachedInputTokens: event.usage.cached_input_tokens,
        reasoningTokens: event.usage.reasoning_output_tokens,
      });
      break;
    case "turn.failed":
      setFailure(event.error?.message ?? "Manager turn failed.");
      break;
    case "error":
      setFailure(event.message);
      break;
    default:
      break;
  }
}

/** The user-facing text of an agent message, or "" when the model signaled silence (a NO_REPLY token
 *  on its own line anywhere in the message). */
export function applyNoReply(text: string): string {
  const silenced = text.split(/\r?\n/).some((line) => line.trim() === NO_REPLY);
  return silenced ? "" : text.trim();
}

function mcpResultText(result: { content: Array<{ type?: string; text?: string }> } | undefined): string | undefined {
  if (!result?.content) return undefined;
  return result.content
    .filter((b) => b.type === "text" && typeof b.text === "string")
    .map((b) => b.text)
    .join("\n");
}

File: src/manager/driver.rs
//! One manager turn over a long-lived backend thread. Port of `src/manager/driver.ts`.
//!
//! Prepend the volatile context header → run the turn → the final agent message is the manager's
//! reply (honoring the `NO_REPLY` sentinel and `ATTACH:` lines); tool calls/reasoning are internal.
//! The driver returns a [`TurnOutcome`]; the app owns delivery + the worker-event reply gate.

use super::backend::{BackendEvent, ManagerBackend, ManagerThread, TurnInput};

/// The sentinel a turn emits to absorb an event without messaging the owner.
pub const NO_REPLY: &str = "NO_REPLY";

/// What a turn produced, for the app to deliver.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TurnOutcome {
    /// Deliver this message (with any image attachments).
    Reply {
        text: String,
        attachments: Vec<String>,
    },
    /// Nothing to say (NO_REPLY, or an empty message).
    Silent,
    /// The turn failed; deliver this owner-facing error text.
    Error(String),
}

/// Drives turns over a backend, owning the lazily-created thread + the resumable session id.
pub struct ManagerDriver {
    backend: Box<dyn ManagerBackend>,
    thread: Option<Box<dyn ManagerThread>>,
    pending_resume: Option<String>,
    current_session: Option<String>,
}

impl ManagerDriver {
    pub fn new(backend: Box<dyn ManagerBackend>) -> Self {
        Self {
            backend,
            thread: None,
            pending_resume: None,
            current_session: None,
        }
    }

    /// The current manager session id, for snapshotting.
    pub fn session_id(&self) -> Option<String> {
        self.current_session.clone()
    }

    /// Seed the resume id from a restored snapshot (before the first turn).
    pub fn adopt_session_id(&mut self, id: Option<String>) {
        self.pending_resume = id.clone();
        self.current_session = id;
    }

    /// `/new`: drop the thread so the next turn starts fresh (memory kept).
    pub fn reset(&mut self) {
        self.thread = None;
        self.pending_resume = None;
        self.current_session = None;
    }

    fn ensure_thread(&mut self) -> &mut Box<dyn ManagerThread> {
        if self.thread.is_none() {
            let resume = self.pending_resume.take();
            self.thread = Some(self.backend.thread(resume));
        }
        match &mut self.thread {
            Some(thread) => thread,
            None => unreachable!("thread was set above"),
        }
    }

    /// Run one turn. `header` is the volatile context header (prepended); `observer` sees every
    /// streamed event (for telemetry/logging).
    pub async fn run_turn(
        &mut self,
        header: &str,
        input: TurnInput,
        observer: &mut (dyn FnMut(&BackendEvent) + Send),
    ) -> TurnOutcome {
        let composed = compose(header, input);

        // Collect events live (forwarding each to the observer), then interpret them afterward.
        let mut events: Vec<BackendEvent> = Vec::new();
        {
            let thread = self.ensure_thread();
            let mut on_event = |ev: BackendEvent| {
                observer(&ev);
                events.push(ev);
            };
            if let Err(err) = thread.run_turn(composed, &mut on_event).await {
                events.push(BackendEvent::Failed(err.to_string()));
            }
        }

        self.capture_session();
        self.build_outcome(events)
    }

    /// Capture the session id once the turn has started one; resume is no longer needed.
    fn capture_session(&mut self) {
        if let Some(id) = self.thread.as_ref().and_then(|t| t.session_id()) {
            self.current_session = Some(id);
            self.pending_resume = None;
        }
    }

    /// Turn the streamed events into an outcome: a failure beats a reply; else the last message.
    fn build_outcome(&self, events: Vec<BackendEvent>) -> TurnOutcome {
        let mut final_reply: Option<String> = None;
        let mut failure: Option<String> = None;
        for ev in events {
            match ev {
                BackendEvent::AgentMessage(t) => final_reply = Some(t),
                BackendEvent::Failed(f) => {
                    failure.get_or_insert(f);
                }
                _ => {}
            }
        }
        if let Some(f) = failure {
            return TurnOutcome::Error(self.backend.format_error(&f));
        }
        final_reply.map_or(TurnOutcome::Silent, |t| finalize_reply(&t))
    }
}

/// Prepend the volatile context header to the turn input.
fn compose(header: &str, input: TurnInput) -> TurnInput {
    let text = if header.is_empty() {
        input.text
    } else {
        format!("{header}\n\n---\n\n{}", input.text)
    };
    TurnInput {
        text,
        image_path: input.image_path,
    }
}

/// Apply NO_REPLY + ATTACH extraction to a raw agent message.
fn finalize_reply(raw: &str) -> TurnOutcome {
    let body = apply_no_reply(raw);
    if body.is_empty() {
        return TurnOutcome::Silent;
    }
    let (text, attachments) = extract_attachments(&body);
    if text.is_empty() && attachments.is_empty() {
        TurnOutcome::Silent
    } else {
        TurnOutcome::Reply { text, attachments }
    }
}

/// The user-facing text, or "" when the model signaled silence (a `NO_REPLY` token on its own line).
pub fn apply_no_reply(text: &str) -> String {
    if text.lines().any(|line| line.trim() == NO_REPLY) {
        String::new()
    } else {
        text.trim().to_string()
    }
}

/// Split a reply into the text the owner reads and the image paths to send. `ATTACH: /path` lines
/// are removed from the text (so NO_REPLY + the delivery gate govern them too).
pub fn extract_attachments(reply: &str) -> (String, Vec<String>) {
    let mut attachments = Vec::new();
    let mut kept = Vec::new();
    for line in reply.lines() {
        let trimmed = line.trim();
        if let Some(rest) = trimmed.strip_prefix("ATTACH:") {
            let path = rest.trim();
            if path.starts_with('/') && !path.contains(char::is_whitespace) {
                attachments.push(path.to_string());
                continue;
            }
        }
        kept.push(line);
    }
    (kept.join("\n").trim().to_string(), attachments)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_reply_silences_whole_message() {
        assert_eq!(apply_no_reply("NO_REPLY"), "");
        assert_eq!(apply_no_reply("thinking...\nNO_REPLY\nmore"), "");
        assert_eq!(apply_no_reply("hello there"), "hello there");
    }

    #[test]
    fn extracts_attachments() {
        let (text, att) = extract_attachments("Here it is:\nATTACH: /tmp/shot.png\ndone");
        assert_eq!(text, "Here it is:\ndone");
        assert_eq!(att, vec!["/tmp/shot.png".to_string()]);
    }

    #[test]
    fn ignores_non_absolute_attach() {
        let (text, att) = extract_attachments("ATTACH: relative.png");
        assert!(att.is_empty());
        assert!(text.contains("ATTACH: relative.png"));
    }

    #[test]
    fn finalize_reply_variants() {
        assert_eq!(
            finalize_reply("hi"),
            TurnOutcome::Reply {
                text: "hi".into(),
                attachments: vec![]
            }
        );
        assert_eq!(finalize_reply("NO_REPLY"), TurnOutcome::Silent);
        assert_eq!(
            finalize_reply("ATTACH: /a.png"),
            TurnOutcome::Reply {
                text: String::new(),
                attachments: vec!["/a.png".into()]
            }
        );
    }
}
